libsql-wal/src/replication/injector.rs

```diff
@@ -0,0 +1,175 @@
+//! The injector is the module in charge of injecting frames into a replica database.
+
+use std::sync::Arc;
+
+use crate::error::Result;
+use crate::io::Io;
+use crate::segment::Frame;
+use crate::shared_wal::SharedWal;
+use crate::transaction::TxGuard;
+
+/// The injector takes frames and injects them in the wal.
+pub struct Injector<'a, IO: Io> {
+    // The wal to which we are injecting
+    wal: Arc<SharedWal<IO>>,
+    buffer: Vec<Box<Frame>>,
+    /// capacity of the frame buffer
+    capacity: usize,
+    tx: TxGuard<'a, IO::File>,
+    max_tx_frame_no: u64,
+}
+
+impl<'a, IO: Io> Injector<'a, IO> {
+    pub fn new(
+        wal: Arc<SharedWal<IO>>,
+        tx: TxGuard<'a, IO::File>,
+        buffer_capacity: usize,
+    ) -> Result<Self> {
+        Ok(Self {
+            wal,
+            buffer: Vec::with_capacity(buffer_capacity),
+            capacity: buffer_capacity,
+            tx,
+            max_tx_frame_no: 0,
+        })
+    }
+
+    pub async fn insert_frame(&mut self, frame: Box<Frame>) -> Result<()> {
+        let size_after = frame.size_after();
+        self.max_tx_frame_no = self.max_tx_frame_no.max(frame.header().frame_no());
+        self.buffer.push(frame);
+
+        if size_after.is_some() || self.capacity == self.buffer.len() {
```

Comment:
should this be `>=`? for the capacity check?

Reply:
That would be safer yeah. But we're injecting frames one by one, so we will hit the ==